meta: add initial `GOVERNANCE.md` (#5040)
jj-vcs/jj: GOVERNANCE.md

Reviewer: nit: line endings are mixed CRLF and LF

# Jujutsu Governance

## Overview

Jujutsu is an open source project, led, maintained and designed for a worldwide
community. Anyone who is interested can join, contribute, and participate in the
decision-making process. This document is intended to help you understand how
you can do that.

## Project roles

There are two broadly defined roles in the Jujutsu project: **Maintainers** and

> **Contributors**.

Reviewer 2: Here are some possible (draft) edits, related to my comment below.

> We greatly appreciate everyone's contributions, and Jujutsu has benefited greatly from people who shared a single idea, change, or a suggestion, without ever becoming a regular contributor. We also want everybody to feel welcome to share their suggestions for the project (as long as you follow the Community Guidelines).
>
> There are two special roles for participants in the Jujutsu projects: **Maintainers** and **Contirbutors**.
>
> The role of the Maintainer is formally defined. These are the people empowered to collectively make final decisions about most aspects of the project. They are expected to take community's input seriously and to aim for the benefit of the entire community.
>
> The role of a Contributor is less formal. In situations where opinions become numerous or contentious, it is acceptable for the maintainers to assign more weight to the voices of the more established Contributors. The Contributors also have more input on who does and does not become a Maintainer, and being a Contributor is usually a prerequisite for becoming a Maintainer.
>

Reviewer 3: > being a Contributor is usually a prerequisite for becoming a Maintainer.

I would even drop "usually" - "Contributor" basically means someone who's a repeat participant in the project, and being a repeat participant is absolutely a mandatory requirement for Maintainership, right?

Otherwise, I like the rephrase - being clear about the distinction between "Contributor" and "random user or single-patch-author" is really helpful. I think it reads as common sense that in a contentious situation we listen to those who have been around for a while more than we listen to passerby.

### Maintainers

**Maintainers** are the people who contribute, review, guide, and collectively
make decisions about the direction and scope of the project (see:
[Decision Making](#decision-making)). Maintainers are elected by a
[voting process](#adding-and-removing-maintainers).

A typical Maintainer is not only someone who has made "large" contributions, but
someone who has shown they are continuously committed to the project and its
community. Some expected responsibilities of maintainers include (but are not
exclusively limited to):

- Displaying a high level of commitment to the project and its community, and
  being a role model for others.
- Writing patches &mdash; a lot of patches, especially "glue code" or "grunt
  work" or general "housekeeping"; fixing bugs, ensuring documentation is always
  high quality, consistent UX design, improving processes, making judgments on
  dependencies, handling security vulnerabilities, and so on and so forth.
- Reviewing code submitted by others &mdash; with an eye to maintainability,
  performance, code quality, and "style" (fitting in with the project).
- Participating in design discussions, especially with regards to architecture

>   or long-term vision.

Reviewer 4: I would explicitly add something about making sure the community remains a welcoming place to users and contributors

Reviewer 3: Yeah, thanks for this comment, I totally agree - all maintainers should be doing this, or at the *very least* should not be having the opposite effect. Would love to have that written down.

This is not an exhaustive list, nor is it intended that every Maintainer does
each and every one of these individual tasks to equal amounts. Rather this is
only a guideline for what Maintainers are expected to conceptually do.

In short, Maintainers are the outwardly visible stewards of the project.

#### Current list of Maintainers

The current list of Maintainers:

- Austin Seipp (@thoughtpolice)
- Ilya Grigoriev (@ilyagr)
- Martin von Zweigbergk (@martinvonz)
- Waleed Khan (@arxanas)
- Yuya Nishihara (@yuja)

### Contributors

We consider contributors to be active participants in the project and community
who are _not_ maintainers. These are people who might:

- Help users by answering questions

> - Participating in lively and respectful discussions across various channels

Reviewer 5: Should we explicitly mention code review either here or in its own bullet point? As currently written, it could be read as saying that only maintainers do code review.

Reviewer 6: We have a "Submit patches or pull requests" bullet. Also mentioning code review in that bullet or a separate bullet would make sense to me.

- Submit high-quality bug reports
- Submit patches or pull requests
- Help with testing and quality assurance
- Submit feedback about planned features, use cases, or bugs

We essentially define them as **people who actively participate in the
project**. Examples of things that would _not_ make you a contributor are:

- Submitting a single bug report and never returning
- Writing blog posts or other evangelism
- Using the software in production
- Forking the project and maintaining your own version
- Writing a third-party tool or add-on

While these are all generally quite valuable, we don't consider these
ongoing contributions to the codebase or project itself, and on their own do not
constitute "active participation".

Contributors and their input are valued and their input is expected to be taken
and respected by Maintainers; not every discussion will result in a change, but
it is intended that every voice will be heard.

## Processes

For the purposes of making decisions across the project, the following processes
are defined.

> ### Decision-Making

Reviewer 2: Overall, I'm not sure whether we want the role of a Contributor to be less well-defined or more well-defined. When writing up the above, I was thinking "less".

OTOH, making it more well-defined would allow us to keep some form of the "temporary voting" process alive as well for larger but [less urgent decisions][^1] (like, say, agreeing on a logo)? I think there is a category of decisions where it's good to get investment and a sense of ownership from the community, or to otherwise rethink them.

Even if we are vague about who exactly is a contributor, we could say that there is a GitHub group that determines eligibility for these kinds of votes. Or, we could try taking everyone's votes and see if any harm comes from that.

[^1]: We could let the maintainers decide whether a particular decision should be decided by themselves or by the contributors. (This also means that we don't necessarily need to change this doc, this could be a policy the maintainers think about later).

Reviewer 3: I'd argue towards less-well-defined to start, because this document can grow, and IMO process and governance works best in "minimum effective dose." And honestly, for voting, I think it's fine to have the artifact around but that temporary voting doc is as stringent as it is because we (gov working group) wanted to make sure we weren't forcing a governance structure onto an unwilling community. For stuff like "which logo should we use?" or "where should we host JJCon next year?", the Maintainers deciding on an ad-hoc system to gather consensus honestly seems fine to me. (In other words, I think that temporary voting doc is actually more complicated than we will almost ever need again.)

Author: Yeah, I agree with Emily that we probably want to keep things a little bit less well defined right now. So I don't think there's anything actionable to do here (unless you have a more specific suggestion)

This document itself is subject to the Decision-Making process by the existing
set of Maintainers.

### Adding and Removing Maintainers

An active Contributor may, at any given time, nominate themselves or another
Contributor to become a Maintainer. This process is purely optional and no
Contributor is expected to do so; however, self-nomination is encouraged for
active participants. A vote and discussion by the existing Maintainers will be
used to decide the outcome.

Note that Contributors should demonstrate a high standard of continuous
participation to become a Maintainer; the upper limit on the number of
Maintainers is practically bounded, and so rejection should be considered as a
real possibility. As the scope of the project changes, this limit may increase,
but it is fundamentally fluid. (If you are unsure, you are free to privately ask
existing Maintainers before self-nominating if there is room.)

A Maintainer may, at any time, cede their responsibility and step down without a
vote.

A Maintainer can be removed by other Maintainers, subject to a vote of at-least
a 2/3rds majority from the existing Maintainer group (excluding the vote of the

> Maintainer in question). This can be due to lack of participation, conduct
> violations, etc. Note that Maintainers are subject to a higher set of behavioral

Reviewer 4: ```suggestion
Maintainer in question). This can be due to lack of participation or conduct violations. Note that Maintainers are subject to a higher set of behavioral
```

Reviewer 7: With this change, it seems to explicitly limit the reasons to remove a Maintainer to the two listed here, and not allow for any others. Is that intentional, or was this just meant to improve the wording? I imagine there could be some other legitimate reason, like lack of alignment with the project goals, or something like that.

Reviewer 4: No, I just found the use of `etc.` as both an abbreviation and to end the sentence confusing, so I tried to edit it out

Reviewer 3: IMO having the list open-ended is a little clearer than having it closed-ended - maybe "lack of participation or conduct violations, among other things"?

# Jujutsu Governance

## Overview

Jujutsu is an open source project, led, maintained and designed for a worldwide
community. Anyone who is interested can join, contribute, and participate in the
decision-making process. This document is intended to help you understand how
you can do that.

## Project roles

There are two broadly defined roles in the Jujutsu project: **Maintainers** and
**Contributors**.

### Maintainers

> At the top of the hierarchy are the **Maintainers**. We consider maintainers to

Reviewer 7: suggestion: I don't think we need to explicitly call out anything "hierarchical", as it seems like somewhat anti-collaborative wording (and, indeed, we added the Contributor role to recognize the valuable contributions of non-Maintainers).

suggestion: I think we should move the actual authority and responsibility up to the top here as the primary definition of a Maintainer, namely that they can conduct votes to resolve conflicts. @nasamuffin already noted below that it didn't belong in the list of Maintainer behaviors.

suggestion: Let's capitalize Maintainer everywhere (assuming that's appropriate?).

being a role model for others.
- Writing patches &mdash; a lot of patches, especially "glue code" or "grunt
work" or general "housekeeping"; fixing bugs, ensuring documentation is always
high quality, consistent UX design, improving processes, making judgments on
dependencies, handling security vulnerabilities, and so on and so forth.
- Reviewing code submitted by others &mdash; with an eye to maintainability,
performance, code quality, and "style" (fitting in with the project).
- Participating in design discussions, especially with regards to architecture
or long-term vision.
- When necessary, participate in a collective voting process to resolve
conflicts and make decisions for which there is no clear answer.

This is not an exhaustive list, nor is it intended that every maintainer does
each and every one of these individual tasks to equal amounts. Rather this is
only a guideline for what maintainers are expected to conceptually do.

In short, Maintainers are the outwardly visible stewards of the project.

#### Current list of maintainers

The current list of maintainers:

- Yuya Nishihara
- Martin von Zweigbergk
- Waleed Khan
- Ilya Grigoriev

> - Austin Seipp

Reviewer 8: Minor Nit: Add the respective GH/Discord usernames here, so its clear how to contact them.

Reviewer 7: + emails, if we think it's relevant

Reviewer: and maybe sort the list?

Author: I think it's probably easiest for now to just put the GitHub usernames. People can just look at our profiles from there to get an email address or method of contact.

- Help users by answering questions
- Participating in lively and respectful discussions across various channels
- Submit high-quality bug reports
- Submit patches or pull requests
- Help with testing and quality assurance
- Submit feedback about planned features, use cases, or bugs

We essentially define them as **people who actively participate in the
project**. Examples of things that would *not* make you a contributor are:

- Submitting a single bug report and never returning
- Writing blog posts or other evangelism
- Using the software in production
- Forking the project and maintaining your own version

While these are all generally quite valuable, they don't directly contribute to
the existing community of users where they are at, and on their own do not
constitute "active participation".

Contributors and their input are valued and their input is expected to be taken
and respected by Maintainers; not every discussion will result in a change, but
it is intended that every voice will be heard.

## Processes

For the purposes of making decisions across the project, the following processes
are defined.

### Decision-Making

> The person proposing a decision to be made (i.e. technical, project direction,
> etc) can offer a proposal, along with a 2-to-4 week deadline for discussion.
> During this time, Maintainers may participate with a vote of:

Reviewer 7: ```suggestion
The person proposing a decision to be made (i.e. technical, project direction,
etc.) can offer a proposal, along with a 2-to-4 week deadline for discussion.
During this time, Maintainers may participate with a vote of:
```

A) Support
B) Reject
C) Abstain

> A final decision requires more than half of the participating votes (i.e.
> excluding abstaining votes.)

Reviewer 7: ```suggestion
A final decision requires more than half of the participating votes (i.e.
excluding abstaining votes).
```

suggestion: When reading this with fresh eyes, I thought it was referring to a kind of 'quorum', i.e. you need to have X non-abstaining votes in order for the vote to be valid. I think that's not what we intended here. I think we should say explicitly something like

> Each Maintainer gets one vote. The total number of "participating votes" is the number of Maintainer votes which are not Abstain. The proposal is accepted when more than half of the participating votes are Support.

And, if necessary, we can give examples of abstention or early-exiting the decision to make it clear what we intended here.

set of Maintainers.

### Adding and Removing Maintainers

An active Contributor may, at any given time, nominate themselves or another
Contributor to become a Maintainer. This process is purely optional and no
Contributor is expected to do so; however, self-nomination is encouraged for
active participants. A vote and discussion by the existing Maintainers will be
used to decide the outcome.

Note that Contributors should demonstrate a high standard of continuous
participation to become a Maintainer; the upper limit on the number of
Maintainers is practically bounded, and so rejection should be considered as a
real possibility. As the scope of the project changes, this limit may increase,
but it is fundamentally fluid. (If you are unsure, you are free to privately ask
existing Maintainers before self-nominating if there is room.)

A Maintainer may, at any time, cede their responsibility and step down without a
vote.

A Maintainer can be removed by other Maintainers, subject to a vote of at-least
a 2/3rds majority from the existing Maintainer group (including the vote of the
Maintainer in question.) This can be due to lack of participation, conduct

> violations, et cetera. Note that Maintainers are subject to a higher set of

Reviewer 7: ```suggestion
violations, etc. Note that Maintainers are subject to a higher set of
```

nit: Why write `et cetera` here but `etc.` elsewhere?

Author: Well, when we initially drafted it we did some parts out of order. Just bad copy :)

# Jujutsu Governance

## Overview

> Jujutsu is an open source project, lead, maintained and designed for a worldwide

Reviewer 3: nit: s/lead/led/

community. Anyone who is interested can join, contribute, and participate in the

> decision making process. This document is intended to help you understand how

Reviewer 6: nit: hyphenate "decision-making" (maybe just my personal preference, but i noticed that wikipedia does it)

you can do that.

## Project roles

There are two broadly defined roles in the Jujutsu project: **Maintainers** and
**Contributors**.

### Maintainers

At the top of the hierarchy are the **Maintainers**. We consider maintainers to
be the people who contribute, review, guide, and collectively make decisions
about the direction and scope of the project.

A typical maintainer is not only someone who has made "large" contributions, but
someone who has shown they are continuously committed to the project and its
community. Some expected responsibilities of maintainers include (but are not
exclusively limited to):

- Displaying a high level of commitment to the project and its community, and
being a role model for others.

> - Writing code &mdash; a lot of code, especially "glue code" or "grunt work";

Reviewer 3: Do we want to say anything about documentation? Is it expected that a documentation-focused maintainer can never write a line of executed code in their time with the project, but still be an expert and guide the usability aspects of the project? Although I'd be quite happy to see that apply to first-party documentation. It doesn't make sense to me, for example, that someone who spills a lot of digital ink about jj on their own personal blog should automagically get to be a maintainer - that doesn't check the "commitment to the health of the project" box for me.

Author: I think that's totally fair. I've reworded this a bit to be clear it's not code, but writing lots of high quality *patches* (which I think is about the right level of abstraction, so to speak.)

fixing bugs, improving processes, making judgments on dependencies, handling
security vulnerabilities, and so on and so forth.
- Reviewing code submitted by others &mdash; with an eye to maintainability,
performance, code quality, and "style" (fitting in with the project).
- Participating in design discussions, especially with regards to architecture
or long-term vision.

> - When necessary, participate in a collective voting process to resolve

Reviewer 3: I worry that this one doesn't quite fit with the others. Voting on conflicts and decisionmaking is a privilege and a duty of a maintainer, not just a thing that someone typically does before being considered for maintainership, right? It might make more sense to call out the duty of the maintainer in a separate list than the "typical behavior of the maintainer" list here.

conflicts and make decisions for which there is no clear answer.

This is not an exhaustive list, nor is it intended that every maintainer does
each and every one of these individual tasks to equal amounts. Rather this is
only a guideline for what maintainers are expected to conceptually do.

In short, Maintainers are the outwardly visible stewards of the project.

#### Current list of maintainers

The current list of maintainers:

- Yuya Nishihara
- Martin von Zweigbergk
- TBD

> TODO: Pending on the initial process of open nominations, decided by Yuya and

Reviewer 6: nit: drop the "on", i think

Martin, more names will be added.

> ### Contributors

Reviewer 6: The Contributor role has no authority associated with it, and I don't think we plan to have a record anywhere of who is considered a Contributor and who is not. So I think this section is mostly about saying that we appreciate people more if they are more active in the project. Should we clarify that here (i.e. that it's not a formal role)? Do people find it useful to have this definition of the informal role?

(Unlike the other nits on this doc, please leave this one open so people can comment, @thoughtpolice)

Reviewer 8: > The Contributor role has no authority associated with it, and I don't think we plan to have a record anywhere of who is considered a Contributor and who is not.

It is useful for a [Rust thanks][thx] like website and to distinguish some longtime contributors from passerby's.  

> So I think this section is mostly about saying that we appreciate people more if they are more active in the project. Should we clarify that here (i.e. that it's not a formal role)?

I think this will work for a few years until the project has a few large components (server, web-frontend, etc), where _knowing_ who to contact for a design or for knowledge behind previous decisions will be useful. 

[thx]: https://thanks.rust-lang.org/

> Do people find it useful to have this definition of the informal role?

IMO, it is useful to have _if_ the project wants to do something with it long-term.

Reviewer 8: Also this doesn't between social moderation roles yet (the current maintainer role in Discord and on GitHub [#2227]), which is something I think should also be defined here.

Reviewer 7: I think it makes sense to write at the top that Contributors are active participants, and that Maintainers are expected to take their feedback into consideration as necessary, but have no formal rights or responsibilities. That should set expectations for how to read the rest of the section.

We consider contributors to be active participants in the project and community
who are *not* maintainers. These are people who might:

- Help users by answering questions
- Participating in lively and respectful discussions across various channels

> - Submit high quality bug reports

Reviewer 6: nit: hyphenate "high-quality"

- Submit patches or pull requests
- Help with testing and quality assurance
- Submit feedback about planned features, use cases, or bugs

Note that we consider Contributors to all be valuable, but we *do* define them

> as people **who actively participate in the project**. Examples of things that

Reviewer 6: This seems to be saying that we consider *non*-Contributors valuable. Is that what you meant? Because the "but" doesn't currently seem to indicate any contradiction, so it's a bit confusing. I'm not sure how to say it in a better way. We could just drop the part saying that we consider them valuable.

> as people **who actively participate in the project**. Examples of things that

Reviewer 6: nit: move "people" inside the `**`? (i first read this as "we define them as people" and then paused a bit in my head, but maybe that's only because i'm reading the markdown source)

would *not* make you a contributor are:

- Submitting a single bug report and never returning
- Writing blog posts or other evangelism
- Using the software in production
- Forking the project and maintaining your own version

While these are all generally quite valuable, they don't directly contribute to

> the existing community of users where they are at, and on their own do not

Reviewer 3: Finding this part pretty hard to understand: "contribute to the existing community of users where they are at"

Maybe can you say in comments in way more words what you meant, and we can try to find a concise way to say it more clearly than this together?

constitute "active participation".

> Contributors and their input is valued and their input is expected to be taken

Reviewer 3: nit: "Contributors and their input are valued"

and respected by Maintainers; not every discussion will result in a change, but
it is intended that every voice will be heard.

## Processes

For the purposes of making decisions across the project, the following processes
are defined.

> ### Decision Making

Reviewer 6: nit: hyphenate

The person proposing a decision to be made (i.e. technical, project direction,

> etc) can offer a proposal, along with a discussion timeline from 2-to-4 weeks.

Reviewer 6: nit: "from" and the hyphenated "2-to-4" seem inconsistent. Drop the hyphens? Or maybe "along with a 2-to-4-week timeline"

Reviewer 3: +1 to the latter suggestion

During this time, Maintainers may participate with a vote of:

A) Support
B) Rejection

> C) Abstinence

Reviewer 6: "Abstention"?

Reviewer 3: Could also try to put them all into the same ...mood? tense? whatever. "Support", "Reject", "Abstain", so they're all verbs

A final decision can be reached upon consensus being achieved, which is defined

> as more than half of the participating votes (i.e. excluding abstaining votes.)

Reviewer 6: Slightly simpler: "A final decision requires more than half of the participating votes (i.e. excluding abstaining votes.)". I mostly wanted to avoid "consensus" and the risk of the reader trying to guess if that means >=50%. We do say that we mean >50%, but the word "consensus" doesn't seem to add anything. I think the line below can also be dropped because it's just the inverse implication from this sentence (i.e. `a => b` implies `!b => !a`).

In the event that a split vote occurs, this would qualify as non-consensus.

In the event that a decision is reached before the proposed timeline, said
proposal can move on and be accepted immediately. In the event no consensus is

> reached, a proposal may be re-submitted later on.

Reviewer 3: Do we also want to make provision for a proposal that was rejected (not just non-consensus) but where new information comes up later on that makes it worth reevaluating?

Reviewer 7: To me, it seems like we don't need to legislate that specifically. Maybe we can move the "In the event no consensus is reached [...]" provision into parentheses to make it clear that there's not a specific process for resubmitting proposals, and you can just resubmit proposals in a reasonable way?

> This document itself is subject to the Decision Making process by the existing

Reviewer 6: nit: hyphenate "Decision-Making" here too

set of Maintainers.

### Adding and Removing Maintainers

An active Contributor may, at any given time, nominate themselves or another
Contributor to become a Maintainer. This process is purely optional and no
Contributor is expected to do so; however, self-nomination is encouraged for
active participants. A vote and discussion by the existing Maintainers will be

> used to decide the outcome.

Reviewer 6: Does this require >50% or >=2/3 of the Maintainers' votes?

Author: With the cleaned-up wording, I think this complaint is handled, unless we think adding new maintainers should have some different threshold. Seems OK to me.

Note that Contributors should demonstrate a high standard of continuous
participation to become a Maintainer; the upper limit on the number of
Maintainers is practically bounded, and so rejection should be considered as a
real possibility. As the scope of the project changes, this limit may increase,
but it is fundamentally fluid. (If you are unsure, you are free to privately ask
existing Maintainers before self-nominating if there is room.)

A Maintainer may, at any time, cede their responsibility and step down without a
vote.

A Maintainer can be removed by other Maintainers, subject to a vote of at-least

> a 2/3rds majority from the existing Maintainer group (including the vote of the

Reviewer 3: What's the reason to include the vote of the person nominated for removal? Usually projects explicitly exclude that person's vote (like in Sociocracy) - I'm curious about the rationale for including it, which seems unusual.

Reviewer 7: IIRC it was somewhat arbitrary. I might have said something vaguely in favor of allowing the person in question to have a say. If it's atypical, then we can adjust it.

I think we also considered what kind of voting thresholds made sense to accommodate the 2/3rds majority. For 5 people, considering whether to include the voted-on Maintainer in their own vote means either —

- if including their vote, then 4/5 of the Maintainers are needed to kick them (which is a unanimous vote from everyone else)
- if not including their vote, then 3/4 of the remaining Maintainers are needed to kick them (which is 3/5 of all Maintainers, allowing up to one dissenter to be overridden)

Actually, writing that down, I think we had been in favor of letting 3/5 of all Maintainers kick one of the other Maintainers. So this provision as written isn't actually consistent with that scenario, and we should probably change it.
